agent/agent.go

```diff
@@ -2106,6 +2106,20 @@ func (a *Agent) AddService(req AddServiceRequest) error {
 func (a *Agent) addServiceLocked(req addServiceLockedRequest) error {
 	req.Service.EnterpriseMeta.Normalize()
 
+	// Must auto-assign the port and default checks (if needed) here to avoid race collisions.
```

Comment:
I guess this needs to be here because we are holding the lock, right? In that case what would happen if we error after, would the port be returned to the pool?

Reply:
yes, if the service does not get successfully registered, the port continues to be available here: https://github.com/hashicorp/consul/blob/main/agent/sidecar_service.go#L155